From 7204cf7c9da9997b1feec5a3a227261073b8ed02 Mon Sep 17 00:00:00 2001
From: czietz <czietz@gmx.net>
Date: Sun, 30 Dec 2018 16:11:22 +0000
Subject: [PATCH] Change the video mode in v_opnwk according to INTIN[0]
 parameter.

Only programs running before the AES are initialized, e.g. from the
AUTO folder, are expected to open the physical workstation. They can
request the video mode while doing so. Particularly, this change
fixes the boot manager XBOOT on STs which boot in low resolution mode.
---
 vdi/vdi_control.c | 19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/vdi/vdi_control.c b/vdi/vdi_control.c
index 7bbecd91..34732bc4 100644
--- a/vdi/vdi_control.c
+++ b/vdi/vdi_control.c
@@ -15,6 +15,7 @@
 #include "vdi_defs.h"
 /* #include "kprint.h" */
 #include "biosbind.h"
+#include "xbiosbind.h"
 #include "../bios/screen.h"
 #include "asm.h"
 #include "string.h"
@@ -390,6 +391,24 @@ void vdi_v_opnwk(Vwk * vwk)
 {
     int i;
     Vwk **p;
+    WORD newrez;
+
+    /*
+     * Programs can request a video mode switch by passing the desired
+     * mode + 2 in INTIN[0]. Falcon-specific modes are currently not
+     * supported.
+     */
+    newrez = INTIN[0] - 2;
+    if (
+        (newrez == ST_LOW) || (newrez == ST_MEDIUM) || (newrez == ST_HIGH)
+#if CONF_WITH_TT_SHIFTER
+        || (newrez == TT_LOW) || (newrez == TT_MEDIUM)
+#endif
+       ) {
+        if (newrez != Getrez()) {
+            Setscreen(-1L, -1L, newrez, 0);
+        }
+    }
 
     /* We need to copy some initial table data from the ROM */
     for (i = 0; i < 12; i++) {
-- 
2.17.0.windows.1

